neurons/miner.py

```diff
@@ -173,51 +177,58 @@ async def forward(
+        # if not self.config.neuron.dont_wandb_log:
+        #     self.wandb.log({"received_indices": synapse.dataset_indices})
 
-        # Encode the dataset
-        encoded_dataset = dataset.map(self.encode, batched=True)
+        # # Encode the dataset
+        # encoded_dataset = dataset.map(self.encode, batched=True)
 
-        # Create a PyTorch DataLoader
-        dataloader = DataLoader(
-            encoded_dataset,
-            batch_size=synapse.batch_size,
-            collate_fn=default_data_collator,
+        # # Create a PyTorch DataLoader
+        # dataloader = DataLoader(
+        #     encoded_dataset,
+        #     batch_size=synapse.batch_size,
+        #     collate_fn=default_data_collator,
+        # )
+
+        # Create Dataloader
+        dataloader = SubsetFalconLoader(
+            batch_size=synapse.batch_size, sequence_length=1024, rows=synapse.dataset_indices
         )
 
-        total_loss = 0
+        n_acc_steps = 0
+        accumulation_steps = self.config.neuron.local_gradient_accumilation_steps_train
 
         # Train data for one epoch
         for step, batch in enumerate(dataloader):
-            input_ids = batch["input_ids"].to(self.device)
-            attention_mask = batch["attention_mask"].to(self.device)
-            labels = input_ids.clone()
-
-            self.opt.zero_grad()
+            inputs = batch.to(self.device)
 
             # Forward pass
-            outputs = self.model(
-                input_ids=input_ids, attention_mask=attention_mask, labels=labels
-            )
-            # Backward pass
-            loss = outputs.loss
-            if not self.config.neuron.dont_wandb_log:
-                self.wandb.log({"loss": loss, "opt_local_epoch": self.opt.local_epoch})
-            total_loss += loss.item()
-
+            outputs = self.model(input_ids=inputs, labels=inputs)
+            
+            # Normalize loss to account for batch accumulation
+            loss = outputs.loss / accumulation_steps  
+            
+            # Backward Pass
             loss.backward()
-            # Adjust gradient
-            self.opt.step()
 
-            bt.logging.info(f"Step {step} Loss: {loss}")
+            if (step + 1) % accumulation_steps == 0:
+                n_acc_steps += 1
+                self.opt.step()         # Adjust gradient
+                self.opt.zero_grad()    # Clear gradients
+                
+                bt.logging.info(f"Step {n_acc_steps} Loss: {outputs.loss.detach().item()}")
+                
+                if not self.config.neuron.dont_wandb_log:
+                    self.wandb.log({"loss": outputs.loss.detach().item(), "opt_local_epoch": self.opt.local_epoch})
+
+            torch.cuda.empty_cache()
 
-        average_loss = total_loss / len(dataloader)
-        synapse.loss = average_loss
+        synapse.loss = outputs.loss.detach().item()
+        synapse.epoch = self.opt.tracker.local_progress.epoch
```

Comment:
Yes exactly so the if statement is to try and detect if the state has changed in the time between the miner's training loop and the validator's validation loop. Because if it has then the validator will be running on a different state and we'll be penalising the miner for having a different loss to that of the validator just because of timing.

Reply:
The average loss was removed because I couldn't get it to work properly late night yesterday whereas normal loss was working fine. I can re-think wether we can add it with a fresh set of eyes today. The main issue will be how do we average a loss that is already averaged across gradient_accumilation_steps.